### PR TITLE
TLS13: CLI: EarlyData: Assign ciphersuite after associated verification in EE

### PR DESCRIPTION
**Pull request by yanrayw**
## Description

**Context**: `tls13 early_data` in `ssl_tls13_client.c`

In `ssl_tls13_postprocess_server_hello`, we assign `handshake->ciphersuite_info->id` to `ssl->session_negotiate->ciphersuite`, this doesn't matter in `non-early_data` situation.  But if client sends `early_data ext` in `client_hello` and receives `early_data ext` in `encrypted_extensions`, the aforementioned assignment makes the [check](https://github.com/Mbed-TLS/mbedtls/blob/development/library/ssl_tls13_client.c#L2193) non-functional. Thus we have to do the assignment after the check in `encrypted_extensions`. This won't cause problems of generating handshake keys as we use `handshake->ciphersuite_info->id` to compute the keys in `ssl_tls13_postprocess_server_hello`.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required
- [x] **backport** not required (tls13 bugfix)
- [x] **tests** not required

### CODE REVIEW

File: library/ssl_tls13_client.c
Comment (yuhaoth) [inline library/ssl_tls13_client.c:1]: >Client receives early data in EE 

What's the meaning? early data extension? Why it is relative ciphersuite?

Comment (xkqian) [reply library/ssl_tls13_client.c:1]: See:         ```
 * RFC 8446 4.2.10
         * In order to accept early data, the server MUST have accepted a PSK
         * cipher suite and selected the first key offered in the client's
         * "pre_shared_key" extension. In addition, it MUST verify that the
         * following values are the same as those associated with the
         * selected PSK:
         * - The TLS version number
         * - The selected cipher suite
         * - The selected ALPN [RFC7301] protocol, if any.
```
    We should check the cipher suite when resume the session and accept early data.    The check is as below:  "    ```
if (handshake->received_extensions & MBEDTLS_SSL_EXT_MASK(EARLY_DATA) &&
        (handshake->selected_identity != 0 ||
         handshake->ciphersuite_info->id !=
         ssl->session_negotiate->ciphersuite)) 
```{". 
But unfortunately, this new check happened in EE, which is after the cipher suite assignment in   ssl_tls13_postprocess_server_hello(), that cause the check useless.

Comment (yuhaoth) [inline library/ssl_tls13_client.c:1]: remove cipher suite check in EE is the right solution.

Comment (xkqian) [reply library/ssl_tls13_client.c:1]: > * ServerHello does not have early data extension, it is sent by EE.
> * EE is first encrypted message, we must set ciphersuite information before that.

I am not trying to question the early data extension in EE.  I and trying to describe why the bug exist. this is what this PR doing.

File: library/ssl_tls13_client.c [hunk 1]
```diff
@@ -2174,24 +2174,8 @@ static int ssl_tls13_process_encrypted_extensions(mbedtls_ssl_context *ssl)
...
-         * selected PSK:
-         * - The TLS version number
-         * - The selected cipher suite
-         * - The selected ALPN [RFC7301] protocol, if any
-         *
-         * We check here that when early data is involved the server
-         * selected the cipher suite associated to the pre-shared key
-         * as it must have.
          */
-        if (handshake->selected_identity != 0 ||
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_client.c:2192]: When parsing the ServerHello we do not know if the server will accept the early data or not thus at this point it is valid for the server to select the first proposed pre-shared key and a ciphersuite that is different from the ciphersuite associated to the selected pre-shared key (only the associated hash has to match at this point, but maybe that currently we accept only the same ciphersuite and that would rather be there that we should improve). Then here we know that the server accepts the early data and it makes sense to me to check the ciphersuite.

Comment (yuhaoth) [reply library/ssl_tls13_client.c:2192]: Agree. I just think about that.  We should check ciphersuite also. But we should check it with proposed ciphersuite. 

And I just checked `mbedtls_ssl_tls13_compute_handshake_transform()` , It is computed base on `handshake->ciphersuite_info`. So the first commit does make sense. 

@yanrayw , please revert to the first commit with some comments

Comment (xkqian) [reply library/ssl_tls13_client.c:2192]: Seems we should revert the title of this PR also. `"TLS13: CLI: EarlyData: Remove unnecessary ciphersuite in EE "`

File: library/ssl_tls13_client.c [hunk 2]
```diff
@@ -2203,6 +2202,20 @@ static int ssl_tls13_process_encrypted_extensions(mbedtls_ssl_context *ssl)
...
+     * proposed pre-shared key with a ciphersuite that is different from the
+     * ciphersuite associated to the selected pre-shared key. To address
+     * aforementioned case, we do associated verification when parsing
+     * early_data ext in EncryptedExtensions. Therefore we have to assign
+     * the ciphersuite in current handshake to session_negotiate later than
+     * the associated verification. This won't impact decryption of
+     * EncryptedExtensions since we compute handshake keys by the ciphersuite
+     * in handshake not via the one in session_negotiate.
+     */
+    ssl->session_negotiate->ciphersuite = handshake->ciphersuite_info->id;
```
Comment (yuhaoth) [inline library/ssl_tls13_client.c:2217]:
```suggestion
    /*
     * Move `session_negotiate->ciphersuite` assign here which after early data cipher suite check.(describe the target)
     *
     * `session_negotiate->ciphersuite` is ciphersuite we proposed in client hello and it is not used for computing handshake_transform. And early data need ciphersuite check. 
     * 
     */
    ssl->session_negotiate->ciphersuite = handshake->ciphersuite_info->id;
```

Comment (yuhaoth) [reply library/ssl_tls13_client.c:2217]: Here we should put the action and why we can do the action. 

above early data cipher check, add reason why we should do the check.

**Review by yuhaoth — CHANGES_REQUESTED** [review]
I think the comments should be split to some paragraph

File: library/ssl_tls13_client.c [hunk 3]
```diff
@@ -2185,9 +2184,14 @@ static int ssl_tls13_process_encrypted_extensions(mbedtls_ssl_context *ssl)
          * - The selected cipher suite
          * - The selected ALPN [RFC7301] protocol, if any
          *
-         * We check here that when early data is involved the server
-         * selected the cipher suite associated to the pre-shared key
-         * as it must have.
+         * When parsing EncryptedExtensions, the client does not know if
```
Comment (yuhaoth) [inline library/ssl_tls13_client.c:2187]: Nitpick comments

Split bellow check and split this comments. 
like bellow

```
/* Why selected_identity == 0 neededs */
if ( handshake->slected_identity) {...}

/* Why ciphersuite check needed */
if (handshake->ciphersuite != session_negotiate->ciphersuite */
```

File: library/ssl_tls13_client.c [hunk 4]
```diff
@@ -2203,6 +2207,18 @@ static int ssl_tls13_process_encrypted_extensions(mbedtls_ssl_context *ssl)
     }
 #endif
 
+    /*
+     * Move `session_negotiate->ciphersuite` assignment here which after
+     * early data cipher suite check when receiving "early_data" extension
+     * in EncryptedExtensions.
```
Comment (yuhaoth) [inline library/ssl_tls13_client.c:2213]:
```suggestion
     * Move `session_negotiate->ciphersuite` assignment here which after
     * early data cipher suite check.
```

File: library/ssl_tls13_client.c [hunk 5]
```diff
@@ -2203,6 +2207,18 @@ static int ssl_tls13_process_encrypted_extensions(mbedtls_ssl_context *ssl)
...
+     *
+     * We compute transform_handshake by the cipher suite chosen from
+     * the server in `handshake`. `session_negotiate->ciphersuite` is the
+     * cipher suite negotiated in previous connection and it is not used for
+     * computing transform_handshake.
```
Comment (yuhaoth) [inline library/ssl_tls13_client.c:2218]:
```suggestion
     * the server in `handshake`. `session_negotiate->ciphersuite` is the
     * cipher suite proposed in this session and not used for
     * computing transform_handshake.
```

**Review by yuhaoth — DISMISSED** [review]
LGTM with some nitpick comments

File: library/ssl_tls13_client.c (hunk 5 above)
Comment (yuhaoth) [reply library/ssl_tls13_client.c:2215]: The value is not come from `previous connection` although they are same in most cases

**Review by yuhaoth — DISMISSED** [review]
LGTM

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: library/ssl_tls13_client.c (hunk 5 above)
Comment (yuhaoth) [reply library/ssl_tls13_client.c:2215]: That's a concept issue. I mean the values are same but they are not same.

In theory, we can send different ciphersuite in resumption session(not early data), that is not forbid by spec(please correct me.)

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: library/ssl_tls13_client.c [hunk 6]
```diff
@@ -2185,9 +2182,15 @@ static int ssl_tls13_process_encrypted_extensions(mbedtls_ssl_context *ssl)
...
-         * We check here that when early data is involved the server
-         * selected the cipher suite associated to the pre-shared key
-         * as it must have.
+         * The server has sent an early data extension in its Encrypted
+         * Extension message thus accepted to receive early data. We
+         * check here that the additional constraints on the handshake
+         * parameters, when early data are exchanged, are met,
+         * namely:
+         * - the selected PSK for the handshake was the first one proposed
+         *   by the client.
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply library/ssl_tls13_client.c:2192]: This looks good to me. Thanks.

**Review by xkqian — APPROVED** [review]
LGTM

The CI failure seems not related with PR.

**Comment by ronald-cron-arm** [comment]
@yuhaoth as you participated to the review of this PR I just wanted to check with you that you agree with its latest version.

**Pull request merged**